Feat/guidelines (#175)
SysBioChalmers/yeast-GEM: .github/CONTRIBUTING.md

    ```
    git checkout devel
    ```
	
5. Create a branch for local development based on the previously checked out branch ([see below](#branching-model) for details on the branching model and how to name your branch):
    ```
    git checkout -b name-of-your-branch
    ```
	
6. Now you can make your changes locally!
    * Always make your changes in Matlab and never directly editing the model files. For loading the model use `loadYeastModel.m`, and for saving the model use `saveYeastModel.m`. The latter will ensure all 3 versions of the model (`.xml`, `.yml` & `.txt`) get updated in the same way.
    * If your changes are minor (e.g. a single chemical formula you wish to correct), you can do it directly from the command line.
    * If your changes are not so small and require several steps, create a script that loads the model, reads data (if applicable), changes the model accordingly, and saves the model back.
    * Each script should start with a commented section describing the script, explaining the parameters, and indicating your name and the date it was written. Existing functions can clarify what style should be used.

>     * Store scripts in the appropriate folder in `/ComplementaryScripts` and data (as `.tsv` files) in the appropriate folder in `/ComplementaryData`. If you think no folder is adequate for your script/data, feel free to create your own folder. Note that binary data such as `.mat` structures or `.xls` tables cannot be stored in the repo (as they cannot be version-controlled, and they increment too much the size of the repo).

Reviewer: You can also just add common binary file formats to the .gitignore, such as `.mat` and similar Matlab files, and `.xlsx` and other Office files. This wouldn't require a text change here though.

Reviewer 2: Good idea, although one can always `git add --force`

Author: @edkerk this is actually already implemented, hence the wording "cannot"
@mihai-sysbio was not aware of `git add --force`, hopefully future PRs don't do it :)

>     * When you are done making changes, review locally your changes with `git diff` or any git client, to make sure you are doing to the model what you intended.

Reviewer: ```suggestion
    * When you are done making changes, review locally your changes with `git diff` or any git client, to make sure you are modifying the model as you intended.
```

Thank you very much for contributing to yeast-GEM!

#### Branching model

* `devel`: Is the branch all pull-requests should be based on.

* `master`: Is only touched by the administrator and is the branch with the tested & reviewed model that is released or ready for the next release.

* `gh-pages`: Is only touched by the administrator and for maintaining the [landing page](http://sysbiochalmers.github.io/yeast-GEM/) of yeast-GEM.

* `{chore, doc, feat, fix, refactor, style}/descriptive-name`: Any other branch created in the model. If you work on a fix, start the branch name with `fix/`, if you work on a feature, start the branch name with `feat/`. Examples: `fix/format_reactions` or `feat/new_algorithms`. [See below](#semantic-commits) for more details on the possible actions you can use.
	
#### Semantic commits

Please use concise descriptive commit messages. Ideally, use semantic commit messages to make it easier to show what you are aiming to do:

`action-object: brief description`
* `action` = {`feat`, `fix`, `refactor`, `style`, `doc`, `chore`, `test`}

> * `object` = {`rxn`, `rxn.annot`, `rxn.prop`, `met`, `met.annot`, `met.prop`, `gene`, `gene.annot`, `comp`, `comp.annot`, `data`}

Reviewer: I don't see that much value in the `object` part, you'd probably want to read the full commit message anyway, and to me this seems to make things more complex. But okay to keep it like this.

Reviewer: Below it is actually mentioned that `object` is optional, maybe mention that here already as well.

Author: done

`object` may or may not be used, and refers to the receiver of the action, i.e. what part of the model/data are you modifying:
* `object`: Used when adding/removing/modifying the whole object (`rxn`, `met`, `gene`, `comp` or `data`).
* `object.annot`: Annotation of the object (e.g. KEGG IDs, EC numbers, etc). Applies for `rxn`, `met`, `gene` or `comp`.
* `object.prop`: Any other property of the object (e.g. formulas, charge, stoich coeffs, LB, UB, etc). Applies for `rxn` or `met`.

Some examples:

|commit|commit message|
|:---:|:---:|
|Add new rxns|`feat-rxn: methanol pathway`|
|Remove a metabolite|`fix-met: duplicated citrate`|
|Add metabolite formula|`feat-met.prop: carbohydrate formulas`|
|Fix rxn stoichiometry|`fix-rxn.prop: complex V stoich coeffs`|
|Update gene IDs|`fix-gene.annot: update IDs from SWISSPROT`|
|Format name of compartment|`style-comp.annot: remove uppercases`|
|Split a rxn in 2|`refactor-rxn: split isomerase in 2 steps`|
|Add some data|`feat-data: metabolomics data`|
|Update documentation of function|`doc: addDBnewRxn.m`|
|Update toolbox|`chore: update RAVEN version`|

> More examples [here](https://github.com/SysBioChalmers/yeast-GEM/commits/master). A more detailed explanation or comments can be left in the commit description.

Reviewer: Shouldn't more detailed explanations be encouraged?

Author: done

* **Compatibility:** First of all, make sure that the model is still compatible with the loading/saving wrappers (`loadYeastModel.m` & `saveYeastModel.m`) and that no errors appear. Check also that [`dependencies.txt`](https://github.com/SysBioChalmers/yeast-GEM/blob/master/ModelFiles/dependencies.txt) does not change in any unexpected ways (e.g. an "unknown" toolbox version). Finally, ensure that the SBML fields `model metaid`, `model id` and `model name` never change, as if they change it would create a conflict in the next release.
* **Documentation:** Every change should be justified with a reference/link/argument. This can be provided as data in `/ComplementaryData`, or directly as a comment in the pull request.
* **Reproducibility:** If there are any added scripts, make sure that if you run them, the model gets updated from how it was in `devel` to how it is in the pull request. For this, you may _locally_ switch to the corresponding branch, replace the `.xml` file with the one from `devel` before the changes, and run the associated scripts. Remember to stash any undesired changes afterwards.

> * **Style:** Ensure that the changes to the model are compliant with the model's rxn/met/gene naming conventions (when unsure, take a look at a similar field in the model). Also, make sure that scripts have a compliant style, and datasets are straight-forward to understand.

Reviewer: Should the naming convention (especially IDs) be specified somewhere? Can IDs be reused for instance?

Author: It should definitively be specified. However, I think that requires a full section and would rather take it on a new PR with more discussion. In the meantime, I have included a mention of `getNewIndex.m` for generating new met/rxn ids in step 6 of "Contributing to the model" (and I will soon work on documenting said function).

* When commenting in the review, please comply with our [code of conduct](https://github.com/SysBioChalmers/yeast-GEM/blob/master/.github/CODE_OF_CONDUCT.md).

> * Avoid vague comments and try to be as explicit as possible (e.g.: _"please change this"_ instead of _"this could be changed"_).

Reviewer: To be honest, the examples that you give here seem vague..

Author: these were not intended to be the whole messages, and I have clarified that in the text

  * Inclusion of a major new formalism in the model.
  * Addition of a plurality of pathways.
* A `patch` release is the most common one and is done when only few things have changed in the model, or there are only changes that have to do with format, such as:
  * Adding a single new annotation field.
  * Fixing some chemical formulas/charges.
  * Updating toolboxes.
  * Re-organization of data
  * Refactoring of code.
  
When releasing, please follow these steps:
  1. Make sure all dependencies in `devel` correspond to the setup from the local computer from which the release will be made. If not, make a single commit in `devel` updating this with a `loadYeastModel`/`saveYeastModel` cycle.
  2. Create a pull request from `devel` to `master`, indicating all new features/fixes/etc. and referencing every previous pull request included. Tip: if any [issue](https://github.com/SysBioChalmers/yeast-GEM/issues) gets solved in the release, write in the pull request description "Fixes #X", where "X" is the issue number. That way the issue will be automatically closed after merge.
  3. Merge at least a day after (having at least one accepted review).

>   4. Switch locally to `master` and update `history.md`, by putting at the top a description with what is included in the new version.

Reviewer: Should new entries in `history.md` not have the same text as the PR to master?

Author: yes it should actually be the same text, I have clarified that now
